### ISSUE TITLE
get changelog from npm

### ISSUE THREAD

**Issue by k-yle** [renovatebot/renovate#9061]
**What would you like Renovate to be able to do?**

To be able to fetch the changelog from the npm package itself, if renovate can't find it on GitHub or GitLab.

If npm sees a `CHANGELOG` file, [it will always publish it](https://docs.npmjs.com/cli/v7/using-npm/developers#:~:text=CHANGELOG). This makes it potentially a more reliable than searching through the repo for tags, releases or changelog-like files.

This had advantages over the current approach because:

- It works even if renovate can't find the source repo
- It works even if renovate guess the release/tag version
- It works if the git release versions don't match the npm versions
- It works for packages in monorepos (see #2926), even if they don't specify `repository.directory` in package.json (like `@aws-amplify/auth`)
- It works for public packages that live in private repos
- It works packages that don't live on GitHub or GitLab

[`typedoc-plugin-markdown`](https://unpkg.com/typedoc-plugin-markdown@3.6.0/CHANGELOG.md) and [`@aws-amplify/auth`](https://unpkg.com/@aws-amplify/auth@3.4.26/CHANGELOG.md) are some examples where the current approach fails, but fetching from NPM would work. 

**Did you already have any implementation ideas?**

If getting the changelog from GitHub or GitLab fails, use 

```handlebars
https://unpkg.com/{{package}}@{{version}}/CHANGELOG.md
```

**Comment by rarkins** [comment]
I think this has enough positives that it's worth doing. Some downsides to note:
1. Releases are immutable, so the CHANGELOG entry for a release will never be added (if missing) or updated (if incorrect or incomplete) after publication
2. We are relying on yet-another-external-service in the form of unpkg
3. It's npm-only, naturally

In my experience CHANGELOG entries often *are* updated post-release.

Given these points, I wouldn't *replace* any of our existing logic with this, but instead supplement it just specifically for the npm case. I think we could use it as a fallback for fetching the changelog file if we can't fetch it through our usual means (e.g. it's on a supported platform, and we are able to locate the file).

What do you think?

**Comment by k-yle** [comment]
Sounds great @rarkins!

As you said, I think it would be valuable to have npm as a fallback source if the better options (GitHub/GitLab) fail. 

> 1. Releases are immutable, so the CHANGELOG entry for a release will never be added (if missing) or updated (if incorrect or incomplete) after publication

That's definitely a drawback, but if this is the last resort then I think showing _something_ is better than _nothing_. And if the package has a CI process to generate a changelog alongside the release then this won't be an issue.

**Comment by rarkins** [comment]
Yes, I agree. There are still a few unknowns to explore here - for example we usually try to avoid datasource-specific logic in lib/workers/* files - but I think this is acceptable to move ahead to development with, if you or someone else is willing.

**Comment by marcodejongh** [comment]
This will also fix changelog generation for packages in public npm, published from a private repo.
I wish there was an alternative to unpkg though, as we also have private npm packages (in our own registry) that we'd like changelog generation for. But just fixing the public usecase would be a huge win.

I'd be willing to help out. Are you suggesting @rarkins that we just add fetch changelog last attempt in the global changelog fetching logic? Which I think is in: lib/workers/pr/changelog/index.ts ?

**Comment by viceice** [comment]
Yes, i think here
https://github.com/renovatebot/renovate/blob/74d7691befa635c1220d8569936273bfb950e748/lib/workers/pr/changelog/index.ts#L39

something like:

```ts
if (!res) {
  res = await sourceUkg.getChangeLogJSON({ ...args, releases });
}
```

**Comment by rarkins** [comment]
A reproduction is still needed which demonstrates an update where the changelog is not found by Renovate today but is available via npm/unpkg
